src/polycap-photon.c

```diff
@@ -0,0 +1,274 @@
+		//define some shell size parameters
+		cap_unita[0] = description->profile->ext[0] / (n_shells);
+		cap_unita[1] = 0.;
+		cap_unitb[0] = cap_unita[0] * cos(M_PI/3.);
+		cap_unitb[1] = cap_unita[0] * sin(M_PI/3.);
+		//convert indexed capillary centre to coordinates
+		capx_0 = i_capx * cap_unita[0] + i_capy * cap_unitb[0];
+		capy_0 = i_capx * cap_unita[1] + i_capy * cap_unitb[1];
+	}
+
+	//define selected capillary axis X and Y coordinates
+	//NOTE: Assuming polycap centre coordinates are X=0,Y=0 with respect to photon->start_coords
+	cap_x = malloc(sizeof(double)*(description->profile->nmax+1));
+	if(cap_x == NULL){
+		printf("Could not allocate cap_x memory.\n");
+		exit(1);
+	}
+	cap_y = malloc(sizeof(double)*(description->profile->nmax+1));
+	if(cap_y == NULL){
+		printf("Could not allocate cap_y memory.\n");
+		exit(1);
+	}
+	
+	for(i=0; i<=description->profile->nmax; i++){
+		cap_x[i] = description->profile->ext[i] * capx_0 / description->profile->ext[0];
+		cap_y[i] = description->profile->ext[i] * capy_0 / description->profile->ext[0];
+	}
+
+	//calculate initial photon weight based on capillary channel effective solid angle.
+	//Mathematically, this is the cos of the angle between photon propagation and polycapillary-to-photonsource axis
+	weight = polycap_scalar(photon->start_direction,central_axis);
+	for(i=0; i<photon->n_energies; i++){
+		photon->weight[i] = photon->weight[i] * weight;
+	}
+
+	
+	//polycap_capil_trace should be ran description->profile->nmax at most,
+	//which means it essentially reflected once every known capillary coordinate
+	for(i=0; i<=description->profile->nmax; i++){ 
+		iesc = polycap_capil_trace(ix, photon, description, cap_x, cap_y);
+		if(iesc != 0){ //as long as iesc = 0 photon is still reflecting in capillary
+		//iesc == -2, which means this photon has reached its final point (weight[0] <1e-4)
+			//in old program a new photon is simulated at this point
+		//alternatively, iesc can be 1 due to not finding intersection point, as the photon reached the end of the capillary
+			break;
+		}
+	}
+
+	if(iesc == -2){
+		return -1; //return -1 if photon did not reach end of capillary
+	} else {
+		return 0; //if photon reached end of capillary, return 0
+	}
+}
+
+//===========================================
+// get exit coordinates
+polycap_vector3 polycap_photon_get_exit_coords(polycap_photon *photon)
+{
+	return photon->exit_coords;
+}
+
+//===========================================
+// get exit direction
+polycap_vector3 polycap_photon_get_exit_direction(polycap_photon *photon)
+{
+	return photon->exit_direction;
+}
+
+//===========================================
+// get exit electric vector
+polycap_vector3 polycap_photon_get_exit_electric_vector(polycap_photon *photon)
+{
+	return photon->exit_electric_vector;
+}
+
+//===========================================
+// free a polycap_photon
+void polycap_photon_free(polycap_photon *photon)
+{
+	polycap_rng_free(photon->rng);
```

Comment:
Ik denk niet dat het slim is om de `rng` te gaan free'en hier: er zullen typisch heel veel fotonen geproduceerd worden, maar slechts 1 rng per thread. Als je gaat free'en moet elk foton zijn eigen, unieke rng krijgen wat de boel onnodig zou vertragen

Reply:
Good catch. Maar hoe werkt dat dan in memory terms? Zorgt dit dan niet voor een memory leak als wel elk foton een eigen photon structure maakt (inherent aan de opbouw met doorgeven van start coordinaten etc), waar die rng een deel van uitmaakt?
Bij het definieren van de photon->rng structure kopieer ik wel gewoon de pointer rng, dus mss lost die memory zichzelf op bij het einde van de omp parallel, waar ik rng in het algemeen free voor elke thread?

Reply:
Je hebt inderdaad geen kopie gemaakt van de rng, enkel het adres ernaar is gekopieerd.

En inderdaad: de rng's moeten enkel gefree'd worden op het einde van de omp parellel region, wat momenteel in orde lijkt te zijn